### PR TITLE
typeanal: add error codes for many errors

### PR DESCRIPTION
**Pull request by hauntsaninja**
I only added codes that I was fairly confident about

### CODE REVIEW

**Comment by hauntsaninja** [comment]
Hmm, mypy self check is fine, but the mypyc build fails. E.g., you can minimise this to:
```
λ git diff | cat            
Alias tip: gd | cat
diff --git a/mypy/message_registry.py b/mypy/message_registry.py
index 1b58f56d8..b6f49baab 100644
--- a/mypy/message_registry.py
+++ b/mypy/message_registry.py
@@ -26,6 +26,7 @@ class ErrorMessage(NamedTuple):
 
 
 # Invalid types
+ASDF: Final = ErrorMessage("yikes", codes.VALID_TYPE)
 INVALID_TYPE_RAW_ENUM_VALUE: Final = "Invalid type: try using Literal[{}.{}] instead?"
 
 # Type checker error message constants

λ mypy --config-file mypy_self_check.ini -p mypy -p mypyc     
Success: no issues found in 264 source files

λ CC=clang MYPYC_OPT_LEVEL=0 python3 setup.py --use-mypyc build_ext --inplace          
mypy/message_registry.py:29: error: Cannot determine type of "VALID_TYPE"
```

File: mypy/typeanal.py [hunk 1]
```diff
@@ -298,15 +300,17 @@ def visit_unbound_type_nonoptional(self, t: UnboundType, defining_literal: bool)
                 )
             if isinstance(sym.node, TypeVarExpr) and tvar_def is not None and self.defining_alias:
                 self.fail(
-                    'Can\'t use bound type variable "{}"'
-                    " to define generic alias".format(t.name),
+                    f'Can\'t use bound type variable "{t.name}" to define generic alias',
                     t,
+                    code=codes.TYPE_VAR,
```
Comment (JukkaL) [inline mypy/typeanal.py:305]: Similar to above, `valid-type` seems better here (or falling back to `misc`).

File: mypy/typeanal.py [hunk 2]
```diff
@@ -322,18 +326,20 @@ def visit_unbound_type_nonoptional(self, t: UnboundType, defining_literal: bool)
                 tvar_def is not None and self.defining_alias
             ):
                 self.fail(
-                    'Can\'t use bound type variable "{}"'
-                    " to define generic alias".format(t.name),
+                    f'Can\'t use bound type variable "{t.name}" to define generic alias',
                     t,
+                    code=codes.TYPE_VAR,
```
Comment (JukkaL) [inline mypy/typeanal.py:331]: Similar to above.

File: mypy/typeanal.py [hunk 3]
```diff
@@ -322,18 +326,20 @@ def visit_unbound_type_nonoptional(self, t: UnboundType, defining_literal: bool)
...
                 )
                 return AnyType(TypeOfAny.from_error)
             if isinstance(sym.node, TypeVarTupleExpr):
                 if tvar_def is None:
-                    self.fail(f'TypeVarTuple "{t.name}" is unbound', t)
+                    self.fail(f'TypeVarTuple "{t.name}" is unbound', t, code=codes.TYPE_VAR)
```
Comment (JukkaL) [inline mypy/typeanal.py:336]: Similar to above.

File: mypy/typeanal.py [hunk 4]
```diff
@@ -1296,7 +1346,9 @@ def bind_function_type_variables(
         defs: list[TypeVarLikeType] = []
         for name, tvar in typevars:
             if not self.tvar_scope.allow_binding(tvar.fullname):
-                self.fail(f'Type variable "{name}" is bound by an outer class', defn)
+                self.fail(
+                    f'Type variable "{name}" is bound by an outer class', defn, code=codes.TYPE_VAR
+                )
```
Comment (JukkaL) [inline mypy/typeanal.py:1351]: Similar to above.

**Review by JukkaL — COMMENTED** [review]
Thanks, looks good overall. Left comments about uses of the `type-var` error code, which didn't feel quite right to me.

**Comment by hauntsaninja** [comment]
Thanks, I had based these off of the one in checker.py. Changed!

**Pull request merged**